### PR TITLE
Add static type analysis with PHPStan

### PR DESCRIPTION
**Pull request by oliverklee**
Add Composer scripts for checking the code with PHPStan (while using
a baseline) and for updating the baseline.

When we fix PHPStan-reported issues (or update the PHPStan configuration),
we can update the baseline bit by bit until all issues are fixed.

### CODE REVIEW

**Comment by oliverklee** [comment]
Sigh. PHPStan requires PHP >= 7.1. @Danon Would it be okay if we raised the PHP version requirement to >= 7.1 (which would mean that - following SemVer - the next release needs to be version 3.0.0)?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -34,7 +34,7 @@ jobs:
   unit-tests:
     name: Unit tests
     runs-on: ubuntu-20.04
-    needs: [php-lint]
+    needs: [ php-lint ]
```
Comment (oliverklee) [inline .github/workflows/build.yml:37]: Note: This is from autoformatting in PhpStorm.

**Comment by oliverklee** [comment]
We could also downgrade PHPStan to version 0.9.x (which still runs on PHP 7.0), but then that version does not have the baseline feature, and it also does not receive any bug fixes. I'd really prefer to use the latest PHPStan version instead.

**Comment by Danon** [comment]
What's with the?
```
"ci:php:stan": "@php ./vendor/bin/phpstan",
```

?

On my local machine, which is Windows with WSL, after I installed the dependencies, in `./vendor/bin/phpstan` I have a bash script. Why running it with `@php`?

**Comment by oliverklee** [comment]
The "@php" means "use Composer's PHP process for this". AFAIK, this is necessary so that (non-WSL) Windows knows how to run the file.

**Comment by Danon** [comment]
@oliverklee Yes, I know what `@php` means, but it won't run a bash script. In the screen, you can see that on my local machine `./vendor/bin/phpstan` is a bash script, not PHP script.

**Comment by oliverklee** [comment]
Oh, I wasn't aware of that. Then indeed the `@php` does not make sense, and I'll remove it.

**Comment by Danon** [comment]
I think the relative path to `./vendor/bin/phpstan` is also invalid, composer complins about `.` being an invalid executable. I think the correct should be
```json
  "scripts": {
    "ci": [
      "@ci:static"
    ],
    "ci:php:stan": "phpstan",
    "ci:static": [
      "@ci:php:stan"
    ],
    "phpstan:baseline": "phpstan --generate-baseline"
  }
```

I would seem that the composer looks into `vendor/bin/` by default. Foe example I used `phpunit` that way.

PS: But now, it looks like `ci:php:stan` doesn't make sense, and just this would be enough:
```json
"scripts": {
  "phpstan": "phpstan"
  "phpstan:baseline": "phpstan --generate-baseline"
}
```

**Comment by oliverklee** [comment]
> But now, it looks like `ci:php:stan`  doesn't make sense, and just this would be enough:

The colon is for grouping commands. Running `composer`  on the https://github.com/MyIntervals/emogrifier project then creates an output like this:

image

If you prefer me to not use any group prefixes in the Composer commands, I can remove them again. Should I?

**Comment by oliverklee** [comment]
I have now implemented a solution that
- does not require contributors to have PHIVE installed
- and does not require raising the minimum PHP version from 7.0 to 7.1

**Comment by oliverklee** [comment]
@Danon Do you still need anything from me so you can merge this PR? (I'd like to reduce the number of open PRs as I rebase all of them after one has been merged, and some PRs also depend on other PRs.)

**Comment by oliverklee** [comment]
@Danon I have updated the PR according to your comments (and also updated to the current PHPStan version). Care to review and merge? Thanks!

**Comment by Danon** [comment]
@oliverklee To be honest, I lost sight of what you're doing here, so I'm just gonna allow it.

**Comment by oliverklee** [comment]
Thanks! :heart:

**Pull request merged**